File: euler-method/euler-method.py
from json import dump

ode_func = lambda t, y: -2 * y + 1


def euler_method(h: float, num_pts: int, y0: float):
    t = [(h * i) for i in range(num_pts)]

    y = [y0] + [0 for i in range(num_pts - 1)]

    for i in range(1, num_pts):
        y[i] = (y[i - 1] + h * ode_func(t[i - 1], y[i - 1])).__round__(5)

    return t, y


if __name__ == "__main__":
    h = 0.1

    num_pts = 6

    y0 = 2

    t, y = euler_method(h, num_pts, y0)

    with open("data-euler.json", "w+", encoding="utf8") as file:
        data = {"y": y, "t": t}
        dump(data, file, indent=2)

File: euler-method/euler-method.lua
function ode_func(t, y)
  return (-2 * y + 1)
end

function euler_method(h, num_pts, y0)
  t = {}
  for i = 0, num_pts - 1, 1 do
    t[i + 1] = h * i
  end

  y = { y0 }
  for i = 2, num_pts, 1 do
    y[i] = y[i - 1] + h * ode_func(t[i - 1], y[i - 1])
  end

  return t, y
end

t, y = euler_method(0.1, 6, 2)

for i = 1, #t, 1 do
  print(t[i], y[i])
end
